src/control/cmd/dmg/main.go

```diff
@@ -62,8 +74,42 @@ func (cmd *ctlInvokerCmd) setInvoker(c control.Invoker) {
 	cmd.ctlInvoker = c
 }
 
-func (cmd *hostListCmd) setHostList(hl []string) {
-	cmd.hostlist = hl
+func (cmd *hostListCmd) getHostList() []string {
+	if cmd.hostlist == nil && !cmd.HostList.Empty() {
+		cmd.hostlist = cmd.HostList.Slice()
+	}
+	return cmd.hostlist
+}
+
+func (cmd *hostListCmd) validateHostList() error {
+	return nil
+}
+
+func (cmd *hostListCmd) setHostList(newList *hostlist.HostSet) {
+	cmd.HostList.Replace(newList)
+}
+
+func (cmd *singleHostCmd) getHostList() []string {
+	if cmd.host == "" {
+		if cmd.HostList.Empty() {
+			cmd.host = "localhost"
+		} else {
+			cmd.host = cmd.HostList.Slice()[0]
+		}
+	}
+	return []string{cmd.host}
+}
+
+func (cmd *singleHostCmd) validateHostList() error {
+	numHosts := cmd.HostList.Count()
+	if numHosts > 1 {
+		return fmt.Errorf("must pass in exactly 1 host (got %d)", numHosts)
+	}
+	return nil
+}
```

Comment:
This is an OK way to do it, but it feels a little funny to make the normal implementation use a stub. A slightly different approach would be to add a new local type alias for `ui.HostSetFlag` and give it a custom `UnmarshalFlag()` method, e.g.

```go
type singleHostFlag ui.HostSetFlag

func (shf *singleHostFlag) UnmarshalFlag(value string) error {
	if err := (*ui.HostSetFlag)(shf).UnmarshalFlag(value); err != nil {
		return err
	}

	if shf.Count() != 1 {
		return errors.New("must specify a single host")
	}

	return nil
}
```

Something to consider, not a required change, though.

Reply:
Good point. I like your suggestion, it feels cleaner to me.